### PR TITLE
Fixes error message that instructs user to .detach()

### LINKED ISSUE #172556: Forgetting to detach() a DTensor graph output that does not need computed gradients leads to bad error message
If you look at this example:
```
import torch
import torch.distributed as dist
from torch.distributed._tensor import DTensor, DeviceMesh, Shard, distribute_tensor
from torch.testing._internal.distributed.fake_pg import FakeStore
# Initialize with fake process group
store = FakeStore()
dist.init_process_group(backend="fake", rank=0, world_size=2, store=store)
# Create a device mesh
device_mesh = DeviceMesh("cuda", [0, 1])
# Create two tensors and convert to DTensors
t1 = torch.randn(4, 4, device="cuda")
dt1 = distribute_tensor(t1, device_mesh, [Shard(0)])
dt1.detach_().requires_grad_()
@torch.compile(backend="aot_eager")
def f(x):
    # This fails
    return x.sin(), x.cos()
    # Replacing with this line works
    # return x.sin(), x.cos().detach()
out1, out2 = f(dt1)
out1.sum().backward()
dist.destroy_process_group()
```

You can see that the example does something inefficient: `f` returns two DTensor outputs, and we only backprop against one of them (`out1`). Both outputs require_grad, though, which means that even after running `out1.sum().backward()`, any activations stored in the graph that were part of `out2`'s compute will stay live in memory (even though, in this example, we never actually cared about backpropping `out2`).

In eager, this just leads to wasted memory. Under compile, though, this becomes a hard error:
```
[rank0]:     AOTDispatchAutograd.process_runtime_tangent(
[rank0]:   File "/data/users/hirsheybar/new2/pytorch/torch/_functorch/_aot_autograd/runtime_wrappers.py", line 2179, in process_runtime_tangent
[rank0]:     raise RuntimeError(
[rank0]: RuntimeError:
[rank0]: During the backward, we encountered a tensor subclass where we guessed its
[rank0]: metadata incorrectly.

[rank0]: Expected metadata: (DTensorSpec(mesh=DeviceMesh((2,), 'cuda', stride=(1,)), placements=(Shard(dim=0),), tensor_meta=TensorMeta(shape=torch.Size([4, 4]), stride=(4, 1), dtype=torch.float32), shard_order=(ShardOrderEntry(tensor_dim=0, mesh_dims=(0,)),)), False), expected type: <class 'torch.distributed.tensor.DTensor'>

[rank0]: Runtime metadata: None, runtime type: <class 'torch.Tensor'>

[rank0]: shape: torch.Size([4, 4])
[rank0]: To fix this, your tensor subclass must implement the dunder method __force_to_same_metadata__.
```

The issue is as follows:

(1) torch.compile traces out a backward graph ahead-of-time, before the user actually calls `.backward()`
(2) in order to do this, we need to create some backward inputs to trace the backward portion of the graph with
(3) we make the assumption that for each forward output that is a DTensor, its corresponding backward input will also be a DTensor
(4) if you call `.backward()` on only a subset of your outputs that require grad, though, then the autograd will replace any "unused" gradients with plain tensors of zeros. So at runtime, the value for `grad_out2` that gets plumbed into the backward graph is a plain tensor of zeros. We had no way of correctly guessing that this might happen at trace time.

This is part of the larger issue of "by tracing the backward ahead of time, we can guess tangent metadata incorrectly". And there are two pretty reasonable workarounds for this case:

(1) just ensure that you detach() any graph outputs that you don't plan on computing gradients for. This will be better for memory usage anyway

(2) if you are willing to compile a single fwd-loss-bwd graph, you will also not run into this problem

The main reason for this github issue is that the error message right now is not very good. Maybe we should consider making it more actionable?

cc @chauhang @penguinwu @wanchaol @tianyu-l @wz337 @XilunWu @d4l3k @pragupta @SherlockNoMad @ppwwyyxx @bobrenjc93

### PR DESCRIPTION
**Pull request by arkadip-maitra**
Fixes #172556 

**The conditional distinguishes two different failure modes that need different advice:**

**Expected subclass, got plain Tensor**: This is the common case where users forget to `.detach() `unused outputs. The fix is to call `.detach()`.

**Both are subclasses, metadata mismatch**: This is a subclass implementation issue. The fix is to implement `__coerce_same_metadata_as_tangent__`.

### CODE REVIEW

File: torch/_functorch/_aot_autograd/runtime_wrappers.py [hunk 1]
```diff
@@ -2235,19 +2235,30 @@ def maybe_coerce(x):
...
-Runtime metadata: {str(runtime_meta)}, runtime type: {str(runtime_type)}
-
-shape: {str(orig_x.shape)}
-To fix this, your tensor subclass must implement the dunder method __force_to_same_metadata__.
-"""
-            )
+Expected a {expected_type.__name__} tangent but got a plain Tensor.
+This happens when a compiled function returns multiple outputs that
+require gradients, but .backward() is only called on some of them.
+To fix: call .detach() on forward outputs you don't need gradients for."""
```
Comment (bdhirsh) [inline torch/_functorch/_aot_autograd/runtime_wrappers.py]: This is fair, although two things I would prefer to do / mention:

(1) The user might not know where to put the detach. Ideally, we can map this tangent to the corresponding forward output, and grab the user stacktrace for the node that created that forward output (which should be a pretty good code indicator of what the user needs to call `.detach()` on)

(2) This error can also happen if you are doing something in your code that causes the compile to degenerately end up with huge numbers of graph breaks, and it can be more profitable to just fix the many graph breaks when this happens instead of hunting down for the right place to call `.detach()`. So we should probably also mention something like:
```
This error is also more likely to occur if your compiled model is suffering from a large number of graph breaks. For more advice on finding and fixing graph breaks, see https://docs.pytorch.org/docs/stable/user_guide/torch_compiler/compile/programming_model.graph_breaks_index.html. For more info about this error, see https://github.com/pytorch/pytorch/issues/172556.
```

Comment (arkadip-maitra) [reply torch/_functorch/_aot_autograd/runtime_wrappers.py]: do u think this is good? i will push the changes if it is good. i did need to make some changes to  the arguments nedded in `AOTDispatchAutograd.process_runtime_tangent`

```
[rank0]: During the backward, we encountered a tensor subclass where we guessed its                                                                                                                           
[rank0]: metadata incorrectly.                                                                                                                                                                                
[rank0]: Expected a DTensor tangent but got a plain Tensor (tangent index: 1).                                                                                                                                
[rank0]: This happens when a compiled function returns multiple outputs that                                                                                                                                  
[rank0]: require gradients, but .backward() is only called on some of them.                                                                                                                                   
[rank0]: To fix: call .detach() on forward outputs you don't need gradients for.                                                                                                                              

[rank0]: The problematic output is: forward output at index 1 (0-indexed)

[rank0]: This error is also more likely to occur if your compiled model is suffering
[rank0]: from a large number of graph breaks. For more advice on finding and fixing
[rank0]: graph breaks, see:
[rank0]: https://docs.pytorch.org/docs/stable/user_guide/torch_compiler/compile/programming_model.graph_breaks_index.html

[rank0]: For more info about this error, see:
[rank0]: https://github.com/pytorch/pytorch/issues/172556
```

Comment (bdhirsh) [reply torch/_functorch/_aot_autograd/runtime_wrappers.py]: > The problematic output is: forward output at index 1 (0-indexed)

I still think this is not very actionable (if there are graph breaks then the user won't know which graph this is coming from, or where in the user code this detach needs to go). So my vote would be:

(1) figure out how to grab the user stack trace for this node (we should be able to map this tangent to a given forward output node from the dynamo graph, which should have a stacktrace in its metadata

(2) also include mention which graph this maps to in tlparse for convenience (e.g. `[1/0]` - this info should be somewhere in the `TracingContext` at compile time, so we might just need to stash it on the `autograd.Function` generated by AOTAutograd.

Comment (arkadip-maitra) [reply torch/_functorch/_aot_autograd/runtime_wrappers.py]: Thanks for the feedback. pushed changes. is this actionable enough?

```
[rank0]: During the backward, we encountered a tensor subclass where we guessed its                                                                                                                           
[rank0]: metadata incorrectly.                                                                                                                                                                                
[rank0]: Expected a DTensor tangent but got a plain Tensor (tangent index: 1).                                                                                                                                
[rank0]: This happens when a compiled function returns multiple outputs that                                                                                                                                  
[rank0]: require gradients, but .backward() is only called on some of them.                                                                                                                                   
[rank0]: To fix: call .detach() on forward outputs you don't need gradients for.                                                                                                                              
                                                                                                                                                                                                              
[rank0]: The problematic output is: forward output at index 1 (0-indexed)                                                                                                                                     
                                                                                                                                                                                                              
[rank0]: This error occurred in compiled graph [0/0].                                                                                                                                                         
                                                                                                                                                                                                              
[rank0]: The forward output was created here:                                                                                                                                                                 
[rank0]:   File "/data/issues/172556.py", line 17, in f                                                                                                                                                       
[rank0]:     return x.sin(), x.cos()                                                                                                                                                                          
                                                                                                                                                                                                              
                                                                                                                                                                                                              
[rank0]: This error is also more likely to occur if your compiled model is suffering                                                                                                                          
[rank0]: from a large number of graph breaks. For more advice on finding and fixing                                                                                                                           
[rank0]: graph breaks, see:                                                                                                                                                                                   
[rank0]: https://docs.pytorch.org/docs/stable/user_guide/torch_compiler/compile/programming_model.graph_breaks_index.html                                                                                     

[rank0]: For more info about this error, see:
[rank0]: https://github.com/pytorch/pytorch/issues/172556
```

File: torch/_functorch/_aot_autograd/graph_capture.py [hunk 2]
```diff
@@ -54,6 +54,46 @@
...
+    if not fw_metadata.traced_tangents_descs:
+        return
+
+    output_node = list(fx_g.graph.nodes)[-1]
+    if output_node.op != "output":
+        return
+
+    all_outputs = output_node.args[0]
+    if not all_outputs:
+        return
```
Comment (bdhirsh) [inline torch/_functorch/_aot_autograd/graph_capture.py]: I don't think we need all of these defensive checks / early returns. The fact that the `fx_g` here is a "a joint graph with 1 output node" is pretty firmly baked into all of the surrounding AOTAutograd code

File: torch/_functorch/_aot_autograd/graph_capture.py [hunk 3]
```diff
@@ -54,6 +54,46 @@
...
+
+    for desc in fw_metadata.traced_tangents_descs:
+        stack_trace = None
+
+        if isinstance(desc, TangentAOTInput):
+            output_desc = desc.output
+            output_idx = None
+
+            if isinstance(output_desc, PlainAOTOutput):
+                output_idx = output_desc.idx
```
Comment (bdhirsh) [inline torch/_functorch/_aot_autograd/graph_capture.py]: I'm a bit worried about what happens when not all of the `desc.output` instances are of type `PlainAOTOutput`. That would mean that we wouldn't actually generate a stacktrace for every tangent, correct? And if `len(fw_metadata.tangent_source_stack_traces) < len(fw_metadata.traced_tangents)`, then you might end up correlating the wrong stacktrace for a given tangent.

Comment (arkadip-maitra) [reply torch/_functorch/_aot_autograd/graph_capture.py]: None will get appended. So the length will be same. Append is outside of the if block

File: torch/_functorch/_aot_autograd/graph_capture.py (hunk 2 above)
Comment (arkadip-maitra) [reply torch/_functorch/_aot_autograd/graph_capture.py]: true. removed em

File: torch/_functorch/_aot_autograd/graph_capture.py [hunk 4]
```diff
@@ -54,6 +54,41 @@
...
+) -> None:
+    from .descriptors import PlainAOTOutput, TangentAOTInput
+
+    if not fw_metadata.traced_tangents_descs:
+        return
+
+    output_node = list(fx_g.graph.nodes)[-1]
+    all_outputs = output_node.args[0]
+
+    stack_traces: list[Optional[str]] = []
```
Comment (aorenste) [inline torch/_functorch/_aot_autograd/graph_capture.py]: nit: `list[str | None]`

File: torch/_functorch/_aot_autograd/graph_capture.py [hunk 5]
```diff
@@ -54,6 +54,41 @@
...
+            output_desc = desc.output
+            output_idx = None
+
+            if isinstance(output_desc, PlainAOTOutput):
+                output_idx = output_desc.idx
+
+            if output_idx is not None and output_idx < len(all_outputs):
+                output_arg = all_outputs[output_idx]
+                if isinstance(output_arg, torch.fx.Node):
+                    stack_trace = output_arg.meta.get("stack_trace", None)
```
Comment (aorenste) [inline torch/_functorch/_aot_autograd/graph_capture.py:82]: nit: This might be easier to read as:
```suggestion
        if isinstance(desc, TangentAOTInput):
            output_desc = desc.output
            if isinstance(output_desc, PlainAOTOutput) and output_desc.idx < len(all_outputs):
                    output_arg = all_outputs[output_desc.idx]
                    if isinstance(output_arg, torch.fx.Node):
                        stack_trace = output_arg.meta.get("stack_trace", None)
```

File: torch/_functorch/_aot_autograd/graph_capture.py [hunk 6]
```diff
@@ -54,6 +54,41 @@
...
+
+        stack_traces.append(stack_trace)
+
+    if any(st is not None for st in stack_traces):
```
Comment (aorenste) [inline torch/_functorch/_aot_autograd/graph_capture.py]: nit: You could avoid this loop with something like:
```
stack_trace, got_one = None, False
...
            stack_trace = output_arg.meta.get(...)
            got_one = True
...
if got_one:
```

File: torch/_functorch/_aot_autograd/runtime_wrappers.py [hunk 7]
```diff
@@ -2070,17 +2070,28 @@ def _backward_prologue_functional(
                 "The grad inputs should be same number as forward output tangents"
             )
 
+        stack_traces = metadata.tangent_source_stack_traces or []
```
Comment (aorenste) [inline torch/_functorch/_aot_autograd/runtime_wrappers.py]: nit: Cheaper to use an empty tuple rather than alloc a list
```suggestion
        stack_traces = metadata.tangent_source_stack_traces or ()
```
(same comment below in the tangent section)

File: torch/_functorch/_aot_autograd/runtime_wrappers.py [hunk 8]
```diff
@@ -2070,17 +2070,28 @@ def _backward_prologue_functional(
...
                 (
                     AOTDispatchAutograd.process_runtime_tangent(
                         t,
                         m,
+                        tangent_idx=idx,
+                        tangent_desc=desc,
+                        compile_id_str=metadata.compile_id_str,
+                        tangent_stack_trace=stack_traces[idx]
+                        if idx < len(stack_traces)
+                        else None,
```
Comment (aorenste) [inline torch/_functorch/_aot_autograd/runtime_wrappers.py]: nit: We expect that if stack_traces is non-empty then it should have an entry for each index, right? So maybe:
```suggestion
                        tangent_stack_trace=stack_traces[idx]
                        if stack_traces
                        else None,
```
(same comment below in the tangent section)

alternatively it might be even more readable by changing the definition of stack_traces to be something like:
```
stack_traces = metadata.tangent_source_stack_traces or itertools.repeat(None, len(tangents))
```
and then you can include it in the `zip`...

File: torch/_functorch/_aot_autograd/schemas.py [hunk 9]
```diff
@@ -511,6 +511,12 @@ class ViewAndMutationMeta:
     # This is populated during graph compilation when stream assignments are made
     mutated_inp_stream_indices: Optional[list[Optional[int]]] = None
 
+    # compile ID string (e.g., "1/0") for error messages
+    compile_id_str: Optional[str] = None
```
Comment (aorenste) [inline torch/_functorch/_aot_autograd/schemas.py]: nit:
```suggestion
    compile_id_str: str | None = None
```

File: torch/_functorch/_aot_autograd/schemas.py [hunk 10]
```diff
@@ -511,6 +511,12 @@ class ViewAndMutationMeta:
...
+
+    # help users identify where to add .detach() in their code
+    tangent_source_stack_traces: Optional[list[Optional[str]]] = None
```
Comment (aorenste) [inline torch/_functorch/_aot_autograd/schemas.py]: nit:
```suggestion
    tangent_source_stack_traces: list[str | None] | None = None
```

File: torch/_functorch/_aot_autograd/runtime_wrappers.py [hunk 11]
```diff
@@ -2352,19 +2378,70 @@ def maybe_coerce(x: torch.Tensor) -> torch.Tensor | None:
         orig_x = x
         x = maybe_coerce(x)
         if x is None:
-            raise RuntimeError(
-                f"""
-During the backward, we encountered a tensor subclass where we guessed its
-metadata incorrectly.
+            expected_subclass_got_plain_tensor = (
```
Comment (aorenste) [inline torch/_functorch/_aot_autograd/runtime_wrappers.py]: This is a pretty big block of pure error handling code. Consider putting it into it's own helper function to make the "real" function easier to read (or maybe even two helpers - one for each value of expected_subclass_got_plain_tensor)

File: torch/_functorch/_aot_autograd/runtime_wrappers.py [hunk 12]
```diff
@@ -2352,19 +2378,70 @@ def maybe_coerce(x: torch.Tensor) -> torch.Tensor | None:
...
+https://github.com/pytorch/pytorch/issues/172556"""
+                )
+            else:
+                raise RuntimeError(
+                    f"""
+During the backward, we encountered a tensor subclass where we guessed its
+metadata incorrectly.
+Expected: {expected_meta} (type {expected_type}),
+got: {runtime_meta} (type {runtime_type}), shape: {orig_x.shape}.
+Your tensor subclass must implement __coerce_same_metadata_as_tangent__."""
```
Comment (aorenste) [inline torch/_functorch/_aot_autograd/runtime_wrappers.py]: nice catch

**Review by aorenste — APPROVED** [review]
Nice. All my comments are nits - feel free to ignore any you disagree with.

File: torch/_functorch/_aot_autograd/runtime_wrappers.py (hunk 11 above)
Comment (arkadip-maitra) [reply torch/_functorch/_aot_autograd/runtime_wrappers.py]: thanks for the suggestion. added a helper so its cleaner

File: torch/_functorch/_aot_autograd/runtime_wrappers.py (hunk 8 above)
Comment (arkadip-maitra) [reply torch/_functorch/_aot_autograd/runtime_wrappers.py]: yes. went with just checking if stack_traces

File: torch/_functorch/_aot_autograd/runtime_wrappers.py [hunk 13]
```diff
@@ -2301,9 +2318,89 @@ def __call__(self, *args: Any, **kwargs: Any) -> Any:
 # This is wrapped in a class just for namespacing purposes
 # No need to make it into an actual CompilerWrapper because it doesn't fit the abstract as cleanly
 class AOTDispatchAutograd:
+    @staticmethod
+    def _raise_tangent_metadata_error(
```
Comment (aorenste) [inline torch/_functorch/_aot_autograd/runtime_wrappers.py:2322]: nit: If this just builds the error it shouldn't be called "raise"...

**Comment by aorenste** [comment]
Approved. The test failure looks unrelated but we should probably rebase to be sure before we override.
Maybe this will do it:
@pytorchbot rebase
